### PR TITLE
add gpu and memory usage in metrics

### PR DESCRIPTION
**Pull request by lxning**
## Description

Please include a summary of the feature or issue being fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes #(issue)
#1452 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Feature/Issue validation/testing

Please describe the tests [UT/IT] that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced. 
Please also list any relevant details for your test configuration.

- [ ] Test A
- [ ] Test B

- Logs
model_metrics.log:
2022-02-20T04:51:12,858 [INFO ] pool-3-thread-1 TS_METRICS - GPUUtilization.Percent:0|#Level:Host,device_id:0|#hostname:ip-172-31-37-20,timestamp:1645332672
2022-02-20T04:51:12,858 [INFO ] pool-3-thread-1 TS_METRICS - GPUMemoryUtilization.Percent:0|#Level:Host,device_id:0|#hostname:ip-172-31-37-20,timestamp:1645332672
2022-02-20T04:51:12,858 [INFO ] pool-3-thread-1 TS_METRICS - GPUMemoryUsed.Megabytes:0|#Level:Host,device_id:0|#hostname:ip-172-31-37-20,timestamp:1645332672

cat ab_report.csv:
Benchmark,Batch size,Batch delay,Workers,Model,Concurrency,Requests,TS failed requests,TS throughput,TS latency P50,TS latency P90,TS latency P99,TS latency mean,TS error rate,Model_p50,Model_p90,Model_p99,predict_mean,handler_time_mean,waiting_time_mean,worker_thread_mean,memory_percentage_mean,gpu_percentage_mean,gpu_memory_percentage_mean,gpu_memory_used_mean
AB,1,100,4,https://torchserve.pytorch.org/mar_files/vgg16.mar,100,10000,0,279.86,354,368,414,357.327,0.0,13.31,13.9,14.21,13.57,13.53,341.0,0.34,5.54,11.25,5.88,920.0

## Checklist:

- [ ] Have you added tests that prove your fix is effective or that this feature works?
- [ ] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

### CODE REVIEW

File: ts/metrics/system_metrics.py [hunk 1]
```diff
@@ -48,14 +49,34 @@ def disk_available():
...
+
+num_of_gpu = -1
+def collect_all(mod, gpu):
     """
     Collect all system metrics.
 
     :param mod:
     :return:
     """
+    global num_of_gpu
```
Comment (msaroufim) [inline ts/metrics/system_metrics.py:78]: can we refactor to remove the global variable?

File: ts/metrics/system_metrics.py [hunk 2]
```diff
@@ -48,14 +49,34 @@ def disk_available():
     data = psutil.disk_usage('/').free / (1024 * 1024 * 1024)  # in GB
     system_metrics.append(Metric('DiskAvailable', data, 'GB', dimension))
 
-
-def collect_all(mod):
+def gpu_utilization():
+    if num_of_gpu <= 0:
+        return
+
+    result = subprocess.run(
```
Comment (msaroufim) [inline ts/metrics/system_metrics.py:56]: We use https://github.com/rossumai/nvgpu elsewhere in the code to avoid having to use shell commands

**Review by msaroufim — APPROVED** [review]

**Review by msaroufim — CHANGES_REQUESTED** [review]
Looks good let's just remove the global variable and clarify a few nits

File: frontend/server/src/main/java/org/pytorch/serve/metrics/MetricCollector.java [hunk 3]
```diff
@@ -32,9 +32,11 @@ public MetricCollector(ConfigManager configManager) {
     public void run() {
         try {
             // Collect System level Metrics
-            String[] args = new String[2];
+            String[] args = new String[4];
             args[0] = configManager.getPythonExecutable();
             args[1] = "ts/metrics/metric_collector.py";
+            args[2] = "--gpu";
```
Comment (msaroufim) [inline frontend/server/src/main/java/org/pytorch/serve/metrics/MetricCollector.java:38]: what will call this code? How do we GPU metrics aren't instrumented for CPU only runs

File: docs/metrics.md [hunk 4]
```diff
@@ -32,6 +32,9 @@ The location of log files and metric files can be configured in the [log4j2.xml]
 | MemoryAvailable | host | MB | memory available on host |
 | MemoryUsed | host | MB | memory used on host |
 | MemoryUtilization | host | percentage | memory utilization on host |
+| GPUUtilizatio | host,device_id | percentage | GPU utilization on host,device_id |
```
Comment (msaroufim) [inline docs/metrics.md:35]: typo here: GPUUtilizatio

**Review by msaroufim — CHANGES_REQUESTED** [review]

File: frontend/server/src/main/java/org/pytorch/serve/metrics/MetricCollector.java (hunk 3 above)
Comment (lxning) [reply frontend/server/src/main/java/org/pytorch/serve/metrics/MetricCollector.java:38]: The action "etric_collector.py" is scheduled by Metrics Manager. Parameter "--gpu" is used to tell backend the host type (ie. CPU or GPU).

**Review by maaquib — APPROVED** [review]

**Pull request merged**